### ISSUE TITLE
An allocator for this device has already been registered for sharing

### ISSUE THREAD

**Issue by skyline75489** [microsoft/onnxruntime-genai#965]
This is seen during testing on iOS. Log link: https://github.com/microsoft/onnxruntime-genai/actions/runs/11248015928/job/31272436617

The line that throws exception in ORT is:

https://github.com/microsoft/onnxruntime/blob/c415991c16b02941f678d77e24d75f4837ee87bc/onnxruntime/core/session/environment.cc#L105

It seems to be caused by calling `CreateAndRegisterAllocator` multiple times.

**Comment by baijumeswani** [comment]
We only [create the globals once](https://github.com/microsoft/onnxruntime-genai/blob/77a88c3884f588ae1d01c71ffac59534e9350c41/src/generators.cpp#L35). I am not sure how the `CreateAndRegisterAllocator` is called multiple times. Is there anything different in iOS compared to other platforms?

@RyanUnderhill any ideas how the `OrtGlobals` constructor could be called multiple times? Is it somehow related to our google test setup?

**Comment by skyline75489** [comment]
It is indeed different on iOS, because an adapter named `xcgtest.mm` is needed to make gtest work in XCTest:

https://github.com/microsoft/onnxruntime-genai/pull/964/files

This changes the init behavior and now `OrtGlobals` will be created multiple times for each test suite.

**Comment by RyanUnderhill** [comment]
Sounds like there's something weird going on with the test wrapper, since globals should only be initialized once.

A single test works fine, it's just that it fails on the second test when it creates the globals a second time and overwrites itself?

**Comment by skyline75489** [comment]
@RyanUnderhill yes. To be precise, the static initializer here gets called for the second time: 

```cpp
std::unique_ptr<OrtGlobals>&
GetOrtGlobals() {
  static auto globals = std::make_unique<OrtGlobals>();
  static auto validate = std::make_unique<ValidateShutdown>();  // Must be after the above line so the destructor runs before the above destructor
  return globals;
}
```

ORT are not happy with it.

**Comment by edgchen1** [comment]
if you need global setup and teardown logic for the tests, gtest has a hook for that. I recently updated ORT's iOS unit tests to use it. previously they weren't cleaning things up properly.

https://github.com/microsoft/onnxruntime/blob/a732f7a4b3bcfc33233c50577a7d75d0eddb2dbf/onnxruntime/test/xctest/xcgtest.mm#L98-L110

**Comment by skyline75489** [comment]
Thanks @edgchen1 ! I'll give it a try

**Comment by skyline75489** [comment]
Sadly due to the lack global initializer in GenAI, I can't balance the calls and bypass the allocator issue. I took a different route and implement the tests in Objective-C, which turns out to be working. I'll close this for now.
